packages/opentelemetry-plugin-xml-http-request/src/xhr.ts

```diff
@@ -0,0 +1,463 @@
+    return (original: OpenFunction): OpenFunction => {
+      const plugin = this;
+      return function patchOpen(
+        this: XMLHttpRequest,
+        method: string,
+        url: string,
+        async?: boolean,
+        user?: string | null,
+        pass?: string | null
+      ): void {
+        plugin._createSpan(this, url, method);
+
+        return original.call(this, method, url, true, user, pass);
+      };
+    };
+  }
+
+  /**
+   * Patches the method send
+   * @private
+   */
+  protected _patchSend() {
+    const plugin = this;
+
+    function endSpan(xhr: XMLHttpRequest, eventName: string) {
+      const xhrMem = plugin._xhrMem.get(xhr);
+      if (!xhrMem) {
+        return;
+      }
+      const callbackToRemoveEvents = xhrMem.callbackToRemoveEvents;
+
+      let resourcesCreatedInTheMiddle = xhrMem.resourcesCreatedInTheMiddle;
+      let entries;
+      if (resourcesCreatedInTheMiddle) {
+        entries = resourcesCreatedInTheMiddle.entries.slice();
+      }
+      if (typeof callbackToRemoveEvents === 'function') {
+        callbackToRemoveEvents();
+      }
+      const currentSpan = xhrMem.span;
+      if (currentSpan) {
+        plugin._findResourceAndAddNetworkEvents(currentSpan, entries);
+        currentSpan.addEvent(eventName);
+
+        const url = currentSpan.attributes[AttributeNames.HTTP_URL];
+        if (typeof url === 'string') {
+          const parsedUrl = parseUrl(url);
+
+          currentSpan.setAttribute(AttributeNames.HTTP_STATUS_CODE, xhr.status);
+          currentSpan.setAttribute(
+            AttributeNames.HTTP_STATUS_TEXT,
+            xhr.statusText
+          );
+          currentSpan.setAttribute(AttributeNames.HTTP_HOST, parsedUrl.host);
+          currentSpan.setAttribute(
+            AttributeNames.HTTP_SCHEME,
+            parsedUrl.protocol.replace(':', '')
+          );
+
+          // @TODO do we want to collect this or it will be collected earlier once only or
+          //    maybe when parent span is not available ?
+          currentSpan.setAttribute(
+            AttributeNames.HTTP_USER_AGENT,
+            navigator.userAgent
+          );
+        }
+
+        currentSpan.end();
+        plugin._tasksCount--;
+      }
+      plugin._clearResources();
+    }
+
+    function onError(this: XMLHttpRequest) {
+      endSpan(this, EventNames.EVENT_ERROR);
+    }
+
+    function onTimeout(this: XMLHttpRequest) {
+      endSpan(this, EventNames.EVENT_TIMEOUT);
+    }
+
+    function onLoad(this: XMLHttpRequest) {
+      if (this.status < 299) {
+        endSpan(this, EventNames.EVENT_LOAD);
+      } else {
+        endSpan(this, EventNames.EVENT_ERROR);
+      }
+    }
+
+    function unregister(xhr: XMLHttpRequest) {
+      xhr.removeEventListener('load', onLoad);
+      xhr.removeEventListener('error', onError);
+      xhr.removeEventListener('timeout', onTimeout);
+      const xhrMem = plugin._xhrMem.get(xhr);
+      if (xhrMem) {
+        delete xhrMem.callbackToRemoveEvents;
```

Comment:
delete is not type safe
```suggestion
        xhrMem.callbackToRemoveEvents = undefined;
```